### PR TITLE
Add pre-commit and use serveral linters to ensure a consistent code style

### PR DESCRIPTION
**Pull request by basnijholt**
I've added pre-commit which runs a bunch of linters at each commit.

This ensures a consistent code-style throughout the project.

### CODE REVIEW

**Comment by mvn23** [comment]
Thanks for the PR, I will have a look over the weekend.

File: pyotgw/protocol.py [hunk 1]
```diff
@@ -154,38 +155,44 @@ def line_received(self, line):
...
         frame = bytes.fromhex(match.group(2))
-        if recvfrom == 'E':
+        if recvfrom == "E":
             _LOGGER.warning("Received erroneous message, ignoring: %s", frame)
             return (None, None, None, None, None)
         msgtype = self._get_msgtype(frame[0])
-        if msgtype in (READ_ACK, WRITE_ACK, READ_DATA, WRITE_DATA):
-            # Some info is best read from the READ/WRITE_DATA messages
+        if msgtype in (v.READ_ACK, v.WRITE_ACK, v.READ_DATA, v.WRITE_DATA):
+            # Some info is best read from the READ/v.WRITE_DATA messages
```
Comment (mvn23) [inline pyotgw/protocol.py]: We can leave `v.` out of comments to keep them shorter and improve readability.

File: pyotgw/protocol.py [hunk 2]
```diff
@@ -208,230 +215,238 @@ def _get_msgtype(self, byte):
...
-    async def _process_msg(self, src, msgtype, msgid, msb, lsb):
+    async def _process_msg(self, src, msgtype, msgid, msb, lsb):  # noqa: C901
         """
         Process message and update status variables where necessary.
         Add status to queue if it was changed in the process.
         """
-        # Ignore output to the thermostat ('A') except MSG_TROVRD,
-        # MSG_TOUTSIDE and MSG_ROVRD as they may contain useful values.
+        # Ignore output to the thermostat ('A') except v.MSG_TROVRD,
+        # v.MSG_TOUTSIDE and v.MSG_ROVRD as they may contain useful values.
```
Comment (mvn23) [inline pyotgw/protocol.py]: See above

File: pyotgw/protocol.py [hunk 3]
```diff
@@ -208,230 +215,238 @@ def _get_msgtype(self, byte):
...
         # Other messages cause issues when overriding values sent to the
         # boiler.
-        if src == 'A' and msgid not in [MSG_TROVRD, MSG_TOUTSIDE,
-                                        MSG_ROVRD]:
+        if src == "A" and msgid not in [v.MSG_TROVRD, v.MSG_TOUTSIDE, v.MSG_ROVRD]:
             return
-        # Ignore upstream MSG_TROVRD if override is active on the gateway.
-        if (src == 'B' and msgid == MSG_TROVRD
-                and self.status.get(DATA_ROOM_SETPOINT_OVRD)):
+        # Ignore upstream v.MSG_TROVRD if override is active on the gateway.
```
Comment (mvn23) [inline pyotgw/protocol.py]: Same

File: pyotgw/pyotgw.py [hunk 4]
```diff
@@ -82,23 +99,36 @@ def __init__(self):
         self._protocol = protocol
         self.loop.create_task(self._protocol.set_update_cb(self._send_report))
         if 0 < inactivity_timeout < 3:
-            _LOGGER.error("Inactivity timeout too low. Should be at least 3 "
-                          "seconds, got %d", inactivity_timeout)
+            _LOGGER.error(
+                "Inactivity timeout too low. Should be at least 3 " "seconds, got %d",
```
Comment (mvn23) [inline pyotgw/pyotgw.py]: No need to break the string in two if it fits on a single line.

File: pyotgw/pyotgw.py [hunk 5]
```diff
@@ -383,30 +416,30 @@ def get_mode(self):
...
-    async def set_mode(self, mode, timeout=OTGW_DEFAULT_TIMEOUT):
+    async def set_mode(self, mode, timeout=v.OTGW_DEFAULT_TIMEOUT):
         """
         Set the operating mode to either "Gateway" mode (:mode: =
-        OTGW_MODE_GATEWAY or 1) or "Monitor" mode (:mode: =
-        OTGW_MODE_MONITOR or 0), or use this method to reset the device
-        (:mode: = OTGW_MODE_RESET).
+        v.OTGW_MODE_GATEWAY or 1) or "Monitor" mode (:mode: =
+        v.OTGW_MODE_MONITOR or 0), or use this method to reset the device
+        (:mode: = v.OTGW_MODE_RESET).
```
Comment (mvn23) [inline pyotgw/pyotgw.py]: `v.` should be kept out of docstrings. It has no meaning outside this file.

File: pyotgw/pyotgw.py [hunk 6]
```diff
@@ -417,9 +450,9 @@ def get_led_mode(self, led_id):
         """
         if not self._connected:
             return
-        return self._protocol.status.get("OTGW_LED_{}".format(led_id))
+        return self._protocol.status.get(f"v.OTGW_LED_{led_id}")
```
Comment (mvn23) [inline pyotgw/pyotgw.py]: This needs testing. It looks like the original code doesn't even work...

File: pyotgw/pyotgw.py [hunk 7]
```diff
@@ -444,12 +477,12 @@ def get_led_mode(self, led_id):
         This method is a coroutine
         """
         if led_id in "ABCDEF" and mode in "RXTBOFHWCEMP":
-            cmd = globals().get("OTGW_CMD_LED_{}".format(led_id))
+            cmd = globals().get(f"v.OTGW_CMD_LED_{led_id}")
```
Comment (mvn23) [inline pyotgw/pyotgw.py]: Same, needs testing.

File: pyotgw/pyotgw.py [hunk 8]
```diff
@@ -444,12 +477,12 @@ def get_led_mode(self, led_id):
...
             status = {}
             ret = await self._wait_for_cmd(cmd, mode, timeout)
             if ret is None:
                 return
-            var = globals().get("OTGW_LED_{}".format(led_id))
+            var = globals().get(f"v.OTGW_LED_{led_id}")
```
Comment (mvn23) [inline pyotgw/pyotgw.py]: Same

File: pyotgw/pyotgw.py [hunk 9]
```diff
@@ -461,9 +494,9 @@ def get_gpio_mode(self, gpio_id):
         """
         if not self._connected:
             return
-        return self._protocol.status.get("OTGW_GPIO_{}".format(gpio_id))
+        return self._protocol.status.get(f"v.OTGW_GPIO_{gpio_id}")
```
Comment (mvn23) [inline pyotgw/pyotgw.py]: Same

File: pyotgw/pyotgw.py [hunk 10]
```diff
@@ -493,18 +526,21 @@ def get_gpio_mode(self, gpio_id):
         if gpio_id in "AB" and mode in range(8):
             if mode == 7 and gpio_id != "B":
                 return None
-            cmd = globals().get("OTGW_CMD_GPIO_{}".format(gpio_id))
+            cmd = globals().get(f"v.OTGW_CMD_GPIO_{gpio_id}")
```
Comment (mvn23) [inline pyotgw/pyotgw.py]: Same

File: pyotgw/pyotgw.py [hunk 11]
```diff
@@ -493,18 +526,21 @@ def get_gpio_mode(self, gpio_id):
...
             status = {}
             ret = await self._wait_for_cmd(cmd, mode, timeout)
             if ret is None:
                 return
             ret = int(ret)
-            var = globals().get("OTGW_GPIO_{}".format(gpio_id))
+            var = globals().get(f"v.OTGW_GPIO_{gpio_id}")
```
Comment (mvn23) [inline pyotgw/pyotgw.py]: Same

File: pyotgw/vars.py [hunk 12]
```diff
@@ -250,126 +250,142 @@
...
-                            "could not be removed because it does not exist "
-                            "in the table."),
-    OVERRUN_ERR: RuntimeError("Overrun Error: The processor was busy and "
-                              "failed to process all received characters.")
+    SYNTAX_ERR: SyntaxError(
+        "Syntax Error: The command contained an "
+        "unexpected character or was incomplete."
+    ),
+    BAD_VALUE: ValueError(
+        "Bad Value: The command contained a data value that " "is not allowed."
```
Comment (mvn23) [inline pyotgw/vars.py]: No need to split the string in 2 if it fits on a single line.

File: pyotgw/vars.py [hunk 13]
```diff
@@ -250,126 +250,142 @@
...
+    ),
+    OUT_OF_RANGE: RuntimeError(
+        "Out of Range: A number was specified outside " "of the allowed range."
```
Comment (mvn23) [inline pyotgw/vars.py]: Same

**Comment by basnijholt** [comment]
Unfortunately, black is uncompromising and doesn't allow you to configure how it handles specific cases. 

https://github.com/psf/black#the-uncompromising-code-formatter

So it's double quotes and limits the line length to 88.

File: pyotgw/protocol.py [hunk 14]
```diff
@@ -45,10 +45,10 @@ def connection_made(self, transport):
         self._cmdq = asyncio.Queue(loop=self.loop)
         self._msgq = asyncio.Queue(loop=self.loop)
         self._updateq = asyncio.Queue(loop=self.loop)
-        self._readbuf = b''
+        self._readbuf = b""
         self._update_cb = None
         self._received_lines = 0
-        self._msg_task = self.loop.create_task(self._process_msgs())
+        self._v.msg_task = self.loop.create_task(self._process_msgs())
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply pyotgw/protocol.py]: I'll fix it. I used Regex to change all of the occurrences but of course it wasn't correct everywhere 😅 I'll fix it.

File: .pre-commit-config.yaml [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+  rev: 19.10b0
+  hooks:
+  - id: black
+- repo: https://gitlab.com/pycqa/flake8
+  rev: 3.7.9
+  hooks:
+  - id: flake8
+    args:
+    - --max-line-length=500
+    - --ignore=E203,E266,E501,W503
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply .pre-commit-config.yaml]: Black (maintained by the Python software fountain) recommends and defaults to 88. Should I change that default to 79?

It's the only thing one is able to configure using black.

[the remark this answers is not in the document]
Comment (mvn23) [reply .pre-commit-config.yaml]: I can live with 88, but 500 seems a bit much

**Comment by mvn23** [comment]
> So it's double quotes and limits the line length to 88.

So be it then :)

File: .pre-commit-config.yaml (hunk 15 above)
[the remark this answers is not in the document]
Comment (basnijholt) [reply .pre-commit-config.yaml]: I changed it to a mimal config that works with `black`.

File: pyotgw/pyotgw.py (hunk 9 above)
Comment (basnijholt) [reply pyotgw/pyotgw.py]: Fixed in https://github.com/mvn23/pyotgw/pull/12/commits/f6ed560b474bbd63057d40757dc13a2988a73c72.

File: pyotgw/pyotgw.py (hunk 10 above)
Comment (basnijholt) [reply pyotgw/pyotgw.py]: Fixed in https://github.com/mvn23/pyotgw/pull/12/commits/f6ed560b474bbd63057d40757dc13a2988a73c72.

File: pyotgw/pyotgw.py (hunk 11 above)
Comment (basnijholt) [reply pyotgw/pyotgw.py]: Fixed in https://github.com/mvn23/pyotgw/pull/12/commits/f6ed560b474bbd63057d40757dc13a2988a73c72.

File: pyotgw/pyotgw.py (hunk 5 above)
Comment (mvn23) [reply pyotgw/pyotgw.py]: Once this is addressed we can merge.

Comment (basnijholt) [reply pyotgw/pyotgw.py]: Done! 👍

**Comment by mvn23** [comment]
Awesome! Thanks again for the PR!

**Pull request merged**